concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+and promise that only a known number of types will implement it.
+
+In Go, *there is no direct support for sum types*.
+The nearest thing possible is to create an interface with an unexported method,
+and make sure only a known number of types in the same package implement it.
+(See https://github.com/BurntSushi/go-sumtype for tooling to help support this;
+the readme also contains good documentation of the concept!)
+
+In Rust, the `enum` keyword (confusingly) also creates sum types.
+Rust's sum types are at least properly closed (unlike Java's and Go's).
+
+#### maybe types
+
+In *many* languages, you may find a type in the standard library called "Maybe",
+or equally frequently, "Optional", or, "Either".
+These are a sum type!
+They're typically explained as "either **none**, or **some** of a thing":
+so, they take the cardinality of a thing and simply "plus one" to it,
+and colloquially call that additional state the "none" state.
+
+For some `Maybe<T>`, the cardinality is "1+|T|".
+
+(This is also a great example for how we can use cardinality counting methods
+to create clear comparisons between differently named but functionally
+identical concepts between programming language communities!)
+
+### product types in the wild
+
+In Java, `class` creates a product type.
+
+In Go, `struct` creates a product type.
+
+In Rust, `struct` creates a product type.
+
+### interfaces in the wild
+
+In Java, the `interface` keyword introduces an interface.
+These interfaces are [*explicit*](#explicit-versus-structural-interfaces) -- a class must use the `implements` keyword to explicitly state that it can satisfy that interface.
+(That the class has sufficient methods to match the interface is then also checked at compile time -- but having the methods is not necessary rather than sufficient for the "implements" relationship to be considered true).
+
+In Go, the `interface` keyword introduces an interface.
+These interfaces are [*structural*](#explicit-versus-structural-interfaces) -- a struct doesn't need to *say* it implements an interface;
+if it has the relevant methods, then you can start using values of that type wherever the interface is required.
+
+In Rust, the `trait` feature combined with the `impl`+`for` keywords is most comparable to the concept of interface in other languages.
+Implementing a trait is [*explicit*](#explicit-versus-structural-interfaces) -- a type must use the `impl`+`for` keywords to explicitly state the methods that satisfy the trait.
+
+### what about numbers?
+
+Numbers are... interesting.
+
+For the most part in computing, we tend to pretend and build abstractions as if
+numbers are [infinite scalars](#infinite-scalars).
+
+Of course, this is rarely true in practice.
+Integers are often implemented in such a way that values that would require beyond 32 (or 64, etc) bits are unrepresentable;
+arguably, this makes them more like a (very (*very*) large) enum than anything else.
```

Comment:
It may be worth mentioning that most languages will provide an infinite-recursive type in the standard library (or the language itself, e.g. Python) for representing arbitrarily large integers.

Reply:
Good idea.  Adding a section for this.

I'm going to include the claim that "these "big" numbers are typically *not interchangable* with the language's built-in number types".

I know that's not *always* true -- there's always an oddball language out there (in this case, iirc, Ruby actually allows bignum shockingly interchangeably with primitives) -- but I think it's *almost* always true, to the point I'm willing to skip any further qualification.  Do you know any other languages that make them interchangable?  Does python?  Are there qualifications required on *how* interchangable?  (If there's a lot of languages that do this, maybe it deserves a note; but if most of those are with a bunch of "well, except $X" hedging, then it tips back to not deserving expansion.)

Reply:
> Do you know any other languages that make them interchangable?

IIRC Erlang has arbitrary sized numbers by default. I have no idea how small numbers are represented internally.

Reply:
Python does, yes (though you can observe when the switch from primitive to bignum occurs using `type()`).

Reply:
I guess my question is... does it work *without caveats*?

I'm not even entirely sure how to qualify this question in a language agnostic way.  But for example, in java, I can think of about a dozen ways where if one tries to mix `Integer` and `BigInteger`, the result is Not What You'd Want and the pitfalls are numerous.

Reply:
@warpfork This is probably not helpful for the document, but yes, as far as I can tell, there are no caveats. I don't even know if there's a way to prevent integer arithmetic on the built-in numeric types from producing oversized ints.

Reply:
I'm going to leave any editing that should result from this discussion as future work.  I don't feel confident enough or have the time to finish researching this deeply enough to make this section better right now.